modules/rostests/apitests/shell32/SHCreateDataObject.cpp

```diff
@@ -0,0 +1,466 @@
+}
+
+#define ok_wstri(x, y) \
+    ok(wcsicmp(x, y) == 0, "Wrong string. Expected '%S', got '%S'\n", y, x)
+
+static void TestHIDA(PVOID pData, SIZE_T Size, LPCWSTR ExpectRoot, LPCWSTR ExpectPath1, LPCWSTR ExpectPath2)
+{
+    LPIDA pida = (LPIDA)pData;
+
+    ok_int(pida->cidl, 2);
+    if (pida->cidl != 2)
+        return;
+
+    WCHAR FolderPath[MAX_PATH], Item1[MAX_PATH], Item2[MAX_PATH];
+    BOOL bRet = SHGetPathFromIDListW(HIDA_GetPIDLFolder(pida), FolderPath);
+    ok_int(bRet, TRUE);
+    if (!bRet)
+        return;
+    ok_wstri(FolderPath, ExpectRoot);
+
+    CComHeapPtr<ITEMIDLIST_ABSOLUTE> pidl1(ILCombine(HIDA_GetPIDLFolder(pida), HIDA_GetPIDLItem(pida, 0)));
+    CComHeapPtr<ITEMIDLIST_ABSOLUTE> pidl2(ILCombine(HIDA_GetPIDLFolder(pida), HIDA_GetPIDLItem(pida, 1)));
+
+    bRet = SHGetPathFromIDListW(pidl1, Item1);
+    ok_int(bRet, TRUE);
+    if (!bRet)
+        return;
+    ok_wstri(Item1, ExpectPath1);
+
+    bRet = SHGetPathFromIDListW(pidl2, Item2);
+    ok_int(bRet, TRUE);
+    if (!bRet)
+        return;
+    ok_wstri(Item2, ExpectPath2);
+}
+
+static void TestHDROP(PVOID pData, SIZE_T Size, LPCWSTR ExpectRoot, LPCWSTR ExpectPath1, LPCWSTR ExpectPath2)
+{
+    DROPFILES* pDropFiles = (DROPFILES*)pData;
+    ok_int(pDropFiles->fWide, TRUE);
+
+    LPCWSTR Expected[2] = { ExpectPath1, ExpectPath2 };
+
+    DWORD offset = pDropFiles->pFiles;
+    UINT Count = 0;
+    for (;;Count++)
+    {
+        LPCWSTR ptr = (LPCWSTR)(((BYTE*)pDropFiles) + offset);
+        if (!*ptr)
+            break;
+
+        if (Count < _countof(Expected))
+            ok_wstri(Expected[Count], ptr);
+
+        offset += (wcslen(ptr) + 1) * sizeof(WCHAR);
+    }
+    ok_int(Count, 2);
+}
+
+static void TestFilenameA(PVOID pData, SIZE_T Size, LPCWSTR ExpectRoot, LPCWSTR ExpectPath1, LPCWSTR ExpectPath2)
+{
+    LPCSTR FirstFile = (LPCSTR)pData;
+    LPWSTR FirstFileW;
+
+    HRESULT hr = SHStrDupA(FirstFile, &FirstFileW);
+    ok_hex(hr, S_OK);
+    if (!SUCCEEDED(hr))
+        return;
+
+    ok_wstri(ExpectPath1, FirstFileW);
+    CoTaskMemFree(FirstFileW);
+}
+
+static void TestFilenameW(PVOID pData, SIZE_T Size, LPCWSTR ExpectRoot, LPCWSTR ExpectPath1, LPCWSTR ExpectPath2)
+{
+    LPCWSTR FirstFile = (LPCWSTR)pData;
+    ok_wstri(ExpectPath1, FirstFile);
+}
+
+
+static void TestDefaultFormat(PCIDLIST_ABSOLUTE pidlFolder, UINT cidl, PCUIDLIST_RELATIVE_ARRAY apidl)
+{
+    CComPtr<IDataObject> spDataObj;
+    HRESULT hr = pSHCreateDataObject(pidlFolder, cidl, apidl, NULL, IID_PPV_ARG(IDataObject, &spDataObj));
+
+    ok_hex(hr, S_OK);
+    if (!SUCCEEDED(hr))
+        return;
+
+    CComPtr<IEnumFORMATETC> pEnumFmt;
+    hr = spDataObj->EnumFormatEtc(DATADIR_GET, &pEnumFmt);
+
+    ok_hex(hr, S_OK);
+    if (!SUCCEEDED(hr))
+        return;
+
+    UINT Expected[4] = {
+        RegisterClipboardFormatA(CFSTR_SHELLIDLISTA),
+        CF_HDROP,
+        RegisterClipboardFormatA(CFSTR_FILENAMEA),
+        RegisterClipboardFormatA("FileNameW"),
```

Comment:
`CFSTR_FILENAMEW`  doesn't work right?

Reply:
Nope, A only works with the A function, and W only with the W function.